src/Analysis/Ast/Impl/Analyzer/PythonAnalyzer.cs

```diff
@@ -14,145 +14,366 @@
+        private readonly IDependencyResolver<ModuleKey, AnalysisEntry> _dependencyResolver;
+        private readonly Dictionary<ModuleKey, AnalysisEntry> _analysisEntries = new Dictionary<ModuleKey, AnalysisEntry>();
+        private readonly DisposeToken _disposeToken = DisposeToken.Create<PythonAnalyzer>();
+        private readonly object _syncObj = new object();
+        private readonly AsyncManualResetEvent _analysisCompleteEvent = new AsyncManualResetEvent();
         private readonly ILogger _log;
+        private readonly int _maxTaskRunning = 8;
+        private int _runningTasks;
+        private int _version;
 
-        public PythonAnalyzer(IServiceManager services, string root) {
+        public PythonAnalyzer(IServiceManager services) {
             _services = services;
             _log = services.GetService<ILogger>();
+            _dependencyResolver = new DependencyResolver<ModuleKey, AnalysisEntry>(new ModuleDependencyFinder(services.GetService<IFileSystem>()));
+            _analysisCompleteEvent.Set();
+        }
+
+        public void Dispose() => _disposeToken.TryMarkDisposed();
+
+        public Task WaitForCompleteAnalysisAsync(CancellationToken cancellationToken = default)
+            => _analysisCompleteEvent.WaitAsync(cancellationToken);
+
+        public void EnqueueDocumentForAnalysis(IPythonModule module, PythonAst ast, int version, CancellationToken cancellationToken)
+            => AnalyzeDocumentAsync(module, ast, version, cancellationToken).DoNotWait();
+
+        public async Task<IDocumentAnalysis> GetAnalysisAsync(IPythonModule module, int waitTime, CancellationToken cancellationToken) {
+            var key = new ModuleKey(module);
+            AnalysisEntry entry;
+            lock (_syncObj) {
+                if (!_analysisEntries.TryGetValue(key, out entry)) {
+                    return new EmptyAnalysis(_services, (IDocument)module);
+                }
+            }
 
-            _dependencyResolver = services.GetService<IDependencyResolver>();
-            if (_dependencyResolver == null) {
-                _dependencyResolver = new DependencyResolver(_services);
-                _services.AddService(_dependencyResolver);
+            using (var timeoutCts = new CancellationTokenSource()) {
+                if (waitTime >= 0 && !Debugger.IsAttached) {
+                    timeoutCts.CancelAfter(waitTime);
+                }
+
+                using (var cts = CancellationTokenSource.CreateLinkedTokenSource(timeoutCts.Token, cancellationToken)) {
+                    var timeoutToken = timeoutCts.Token;
+                    while (!timeoutToken.IsCancellationRequested) {
+                        try {
+                            var analysis = await entry.GetAnalysisAsync(cts.Token);
+                            lock (_syncObj) {
+                                if (entry.Version == analysis.Version) {
+                                    return analysis;
```

Comment:
So it goes in and out of lock. Do we care if `entry` changes in the dictionary between we fetched it and when we reached `entry.Version == analysis.Version` ?

Reply:
Hm, I can't find where entry is changed outside of the lock.

Reply:
I mean code fetches entry inside lock, then goes out. At this time theoretically entry can get replaced in the dictionary with something else. Or not. Just checking.